ttnn/cpp/ttnn/tensor/tensor_impl.cpp

```diff
@@ -840,6 +847,214 @@ Tensor to_device<bfloat8_b>(
     return to_device<uint32_t>(tensor, target_device, memory_config, cq_id, sub_device_ids);
 }
 
+// ======================================================================================
+//     Helpers for converting between logical <-> physical data with full tensor spec
+// ======================================================================================
+Size flatten_to_2D(const ttnn::SimpleShape& shape) {
+    const int rank = static_cast<int>(shape.rank());
+
+    size_t width = 1;
+    size_t height = 1;
+
+    // Iterate dims in reverse order
+    // Even tensor of rank 0 or 1
+    for (int i = -1; i >= -rank; --i) {
+        auto& dim = i == -1 ? width : height;
+        dim *= shape[i];
+    }
+
+    Size size{height, width};
+
+    return size;
+}
+
+std::array<size_t, 4> compute_shard_spec(const Size& shape, const Size& shard_shape) {
+    const auto num_shards_height = tt::div_up(shape.height(), shard_shape.height());
+    const auto last_shard_height =
+        shape.height() % shard_shape.height() > 0 ? shape.height() % shard_shape.height() : shard_shape.height();
+    const auto num_shards_width = tt::div_up(shape.width(), shard_shape.width());
+    const auto last_shard_width =
+        shape.width() % shard_shape.width() > 0 ? shape.width() % shard_shape.width() : shard_shape.width();
+
+    return {num_shards_height, last_shard_height, num_shards_width, last_shard_width};
+};
+
+// TODO: Remove when we generalize interleaved and sharded; when we do, directly get from TensorLayout
+std::array<Size, 2> get_logical_and_physical_shard_shapes(const TensorSpec& tensor_spec) {
+    if (tensor_spec.memory_config().is_sharded()) {
+        return {
+            tensor_spec.tensor_layout().get_logical_shard_shape(),
+            tensor_spec.tensor_layout().get_physical_shard_shape()};
+    }
+
+    const auto& logical_shape = tensor_spec.logical_shape();
+    Size logical_shard_shape{logical_shape[-2], logical_shape[-1]};
+    auto physical_shard_shape = logical_shard_shape;
+    if (tensor_spec.layout() == Layout::TILE) {
+        const auto& tile = tensor_spec.tile();
+        auto physical_shard_height = tt::round_up(logical_shard_shape.height(), tile.get_height());
+        auto physical_shard_width = tt::round_up(logical_shard_shape.width(), tile.get_width());
+        physical_shard_shape = Size{physical_shard_height, physical_shard_width};
+    }
+    return {logical_shard_shape, physical_shard_shape};
+}
+
+using LogicalPhysicalIdxPairs = std::vector<std::array<size_t, 2>>;
```

Comment:
User a pair here?

Reply:
Why if the type is the same?

Reply:
E.g. you can use structured bindings in the `for` below. But also - if you are intending to express "pair of ids", why not just use a pair directly?